pkg/detector/preemption.go

```diff
@@ -0,0 +1,181 @@
+	"github.com/karmada-io/karmada/pkg/features"
+	"github.com/karmada-io/karmada/pkg/util"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+	"github.com/karmada-io/karmada/pkg/util/names"
+)
+
+// preemptionEnabled checks if preemption is enabled.
+func preemptionEnabled(preemption policyv1alpha1.PreemptionBehavior) bool {
+	if preemption != policyv1alpha1.PreemptAlways {
+		return false
+	}
+	if !features.FeatureGate.Enabled(features.PolicyPreemption) {
+		klog.Warningf("Cannot handle the preemption process because feature gate %q is not enabled.", features.PolicyPreemption)
+		return false
+	}
+	return true
+}
+
+// handleClusterPropagationPolicyPreemption handles the preemption process of PropagationPolicy.
+// If preemption is not enabled, the process will be aborted.
+// The preemption rule: high-priority PP > low-priority PP > CPP.
+func (d *ResourceDetector) handlePropagationPolicyPreemption(policy *policyv1alpha1.PropagationPolicy) error {
+	if preemptionEnabled(policy.Spec.Preemption) {
+		return nil
+	}
+
+	var errs []error
+	for _, rs := range policy.Spec.ResourceSelectors {
+		resourceTemplate, err := d.fetchResourceTemplate(rs)
+		if err != nil || resourceTemplate == nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		errs = append(errs, d.preemptPropagationPolicy(resourceTemplate, policy))
+		// TODO(whitewindmills): PP preempts CPP.
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+// handleClusterPropagationPolicyPreemption handles the preemption process of ClusterPropagationPolicy.
+// If preemption is not enabled, the process will be aborted.
+// The preemption rule: high-priority CPP > low-priority CPP.
+func (d *ResourceDetector) handleClusterPropagationPolicyPreemption(policy *policyv1alpha1.ClusterPropagationPolicy) error {
+	if preemptionEnabled(policy.Spec.Preemption) {
+		return nil
+	}
+
+	var errs []error
+	for _, rs := range policy.Spec.ResourceSelectors {
+		resourceTemplate, err := d.fetchResourceTemplate(rs)
+		if err != nil || resourceTemplate == nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		errs = append(errs, d.preemptClusterPropagationPolicy(resourceTemplate, policy))
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+// preemptPropagationPolicy preempts resource template that is claimed by PropagationPolicy.
+func (d *ResourceDetector) preemptPropagationPolicy(resourceTemplate *unstructured.Unstructured, policy *policyv1alpha1.PropagationPolicy) error {
+	rtLabels := resourceTemplate.GetLabels()
+	claimedPolicyNamespace := util.GetLabelValue(rtLabels, policyv1alpha1.PropagationPolicyNamespaceLabel)
+	claimedPolicyName := util.GetLabelValue(rtLabels, policyv1alpha1.PropagationPolicyNameLabel)
+	if claimedPolicyName == "" || claimedPolicyNamespace == "" {
+		return nil
+	}
+	// resource template has been claimed by policy itself.
+	if claimedPolicyNamespace == policy.Namespace && claimedPolicyName == policy.Name {
+		return nil
```

Comment:
Here, should we consider the resource comes  from different groups?

Reply:
sorry, I don't get you.